editor/code_editor.cpp

```diff
@@ -1498,87 +1488,64 @@ void CodeTextEditor::move_lines_down() {
 	text_editor->queue_redraw();
 }
 
-void CodeTextEditor::_delete_line(int p_line, int p_caret) {
-	// this is currently intended to be called within delete_lines()
-	// so `begin_complex_operation` is omitted here
-	text_editor->set_line(p_line, "");
-	if (p_line == 0 && text_editor->get_line_count() > 1) {
-		text_editor->set_caret_line(1, p_caret == 0, true, 0, p_caret);
-		text_editor->set_caret_column(0, p_caret == 0, p_caret);
-	}
-	text_editor->backspace(p_caret);
-	if (p_line < text_editor->get_line_count()) {
-		text_editor->unfold_line(p_line);
-	}
-	text_editor->set_caret_line(p_line, p_caret == 0, true, 0, p_caret);
-}
-
 void CodeTextEditor::delete_lines() {
 	text_editor->begin_complex_operation();
 
-	Vector<int> carets_to_remove;
-
-	Vector<int> caret_edit_order = text_editor->get_caret_index_edit_order();
-	for (int i = 0; i < caret_edit_order.size(); i++) {
-		int c = caret_edit_order[i];
-		int cl = text_editor->get_caret_line(c);
-
-		bool swaped_caret = false;
-		for (int j = i + 1; j < caret_edit_order.size(); j++) {
-			if (text_editor->has_selection(caret_edit_order[j])) {
-				if (text_editor->get_selection_from_line() == cl) {
-					carets_to_remove.push_back(caret_edit_order[j]);
-					continue;
-				}
-
-				if (text_editor->get_selection_to_line() == cl) {
-					if (text_editor->has_selection(c)) {
-						if (text_editor->get_selection_to_line(c) != cl) {
-							text_editor->select(cl + 1, 0, text_editor->get_selection_to_line(c), text_editor->get_selection_to_column(c), c);
-							break;
-						}
-					}
-
-					carets_to_remove.push_back(c);
-					i = j - 1;
-					swaped_caret = true;
-					break;
-				}
-				break;
+	Vector<int> lines = _get_affected_lines();
+	for (int c = 0; c < text_editor->get_caret_count(); c++) {
+		text_editor->deselect(c);
+	}
+	for (const int &line : lines) {
+		// Save caret positions from the line and then remove the carets.
+		Vector<int> caret_columns;
+		Vector<int> carets_to_remove;
+		int indent = text_editor->get_indent_size();
+		for (int c = 0; c < text_editor->get_caret_count(); c++) {
+			if (line == text_editor->get_caret_line(c)) {
+				int visual_column = text_editor->get_caret_column(c);
+				visual_column += MIN(visual_column, text_editor->get_indent_level(line) / indent) * (indent - 1);
+				print_line("From column:", text_editor->get_caret_column(c), "visual column:", visual_column, "indent:", text_editor->get_indent_level(line), "*", text_editor->get_indent_size());
+				caret_columns.append(visual_column);
+				carets_to_remove.append(c);
```

Comment:
Storing carets temporarily somewhere is quite hacky solution, and we need to take care that they won't overlap since backspace operation merges overlapping carets. I feel that removing and reinserting carets it the safest way to do this, removes the risk of something unexpected happening in the textedit not only now, but in future too, if something were to change on that end.

I tried this just in case it would cure the undo problem, but somehow it made it worse, now even the main caret doesn't get reset.

As for optimizing, I think that can be safely postponed until somebody actually reports a performance problem. I think it would add unnecessary code complexity to not hot code. Hence my not very optimized, but clear steps.

One possible solution would be to add `delete_line` to TextEdit, which could then handle things more gracefully. It does seem like a strange omission taken that we have `insert_line_at` and even `swap_lines` 🤷‍♂️ 
Speaking of `swap_lines`, that could hold another option - if we don't remove this line but swap with the next line, carets would already be at the right line. Then we could safely remove the next line. I quickly tested it but it yielded some strange things so I need to give it tomorrow a second look - it's already late here.

Reply:
That's good point, `remove_text` is now exposed, so might be worth using that instead, though that will not make any caret adjustments.

Reply:
Oh, I didn't know we had `remove_text`! Now it's progressing nicely. I got undo to work correctly. There's still some edge cases that needs to be ironed out, I hope I'll have some more time tomorrow to finish it.